internal/webhooks/patch_validation.go

```diff
@@ -0,0 +1,360 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package webhooks
+
+import (
+	"encoding/json"
+	"fmt"
+	"strconv"
+	"strings"
+	"text/template"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+)
+
+// validatePatches returns errors if the Patches in the ClusterClass violate any validation rules.
+func validatePatches(clusterClass *clusterv1.ClusterClass) field.ErrorList {
+	var allErrs field.ErrorList
+	names := sets.String{}
+	for i, patch := range clusterClass.Spec.Patches {
+		allErrs = append(
+			allErrs,
+			validatePatch(patch, names, clusterClass, field.NewPath("spec", "patches").Index(i))...,
+		)
+		names.Insert(patch.Name)
+	}
+	return allErrs
+}
+
+func validatePatch(patch clusterv1.ClusterClassPatch, names sets.String, clusterClass *clusterv1.ClusterClass, path *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	allErrs = append(allErrs,
+		validatePatchName(patch, names, path)...,
+	)
+	allErrs = append(allErrs,
+		validatePatchDefinitions(patch, clusterClass, path)...,
+	)
+	return allErrs
+}
+
+func validatePatchName(patch clusterv1.ClusterClassPatch, names sets.String, path *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	if patch.Name == "" {
+		allErrs = append(allErrs,
+			field.Required(
+				path.Child("name"),
+				"patch name must be defined",
+			),
+		)
+	}
+
+	if names.Has(patch.Name) {
+		allErrs = append(allErrs,
+			field.Invalid(
+				path.Child("name"),
+				patch.Name,
+				fmt.Sprintf("patch names must be unique. Patch with name %q is defined more than once.", patch.Name),
```

Comment:
```suggestion
				fmt.Sprintf("patch names must be unique. Patch with name %q is defined more than once", patch.Name),
```
nit, trailing dot in error message (or at least the string which will be the end of the error message)

Reply:
Intellij likes to hide error strings from me!